feat(xo-server-auth-ldap): import LDAP groups

pdonias opened the pull request:
See #1884 (First step of the issue)

When a user logs into XO using LDAP:
- Create a XO user if it doesn't already exist
- Mark it as being provided by LDAP and bind it to the LDAP user with an ID
- If group synchronization is enabled:
  - Fetch all the LDAP groups based on the Base and Filter (plugin config)
  - Create and delete the corresponding XO groups based on the LDAP groups that
    were found
  - Add and remove the XO users from the XO groups based on the LDAP data

### Check list

> Check if done, if not relevant leave unchecked.

- [x] PR reference the relevant issue (e.g. `Fixes #007` or `See xoa-support#42`)
- [ ] if UI changes, a screenshot has been added to the PR
- [ ] documentation updated (in another PR)
- `CHANGELOG.unreleased.md`:
  - [x] enhancement/bug fix entry added
  - [x] list of packages to release updated (`${name} v${new version}`)
- **I have tested added/updated features** (and impacted code)
  - [ ] unit tests (e.g. [`cron/parse.spec.js`](https://github.com/vatesfr/xen-orchestra/blob/b24400b21de1ebafa1099c56bac1de5c988d9202/%40xen-orchestra/cron/src/parse.spec.js))
  - [ ] if `xo-server` API changes, the corresponding test has been added to/updated on [`xo-server-test`](https://github.com/vatesfr/xen-orchestra/tree/master/packages/xo-server-test)
  - [x] at least manual testing

### Process

1. create a PR as soon as possible
1. mark it as `WiP:` (Work in Progress) if not ready to be merged
1. when you want a review, add a reviewer (and only one)
1. if necessary, update your PR, and re- add a reviewer

From [_the Four Agreements_](https://en.wikipedia.org/wiki/Don_Miguel_Ruiz#The_Four_Agreements):

1. Be impeccable with your word.
1. Don't take anything personally.
1. Don't make assumptions.
1. Always do your best.

pdonias committed:
feat(xo-server-auth-ldap): import LDAP groups

See #1884

When a user logs into XO using LDAP:
- Create a XO user if it doesn't already exist
- Mark it as being provided by LDAP and bind it to the LDAP user with an ID
- If group synchronization is enabled:
  - Fetch all the LDAP groups based on the Base and Filter (plugin config)
  - Create and delete the corresponding XO groups based on the LDAP groups that
    were found
  - Add and remove the XO users from the XO groups based on the LDAP data

julien-f commented on packages/xo-server-auth-ldap/src/index.js:
```diff
@@ -250,6 +340,161 @@ class AuthLdap {
+    const ldapId = ldapUser[idAttribute]
+    const xoUsers = await this._xo.getAllUsers()
+    let xoUser = xoUsers.find(
+      xoUser => xoUser.authProviders?.ldap?.id === ldapId
```
What if one day we wish to have multiple LDAP providers?

julien-f commented on packages/xo-server-auth-ldap/src/index.js:
```diff
@@ -250,6 +340,161 @@ class AuthLdap {
+    const { merge, idAttribute } = this._usersConfig
+
+    const ldapId = ldapUser[idAttribute]
+    const xoUsers = await this._xo.getAllUsers()
```
What if there are thousands of users?

julien-f commented on packages/xo-server-auth-ldap/src/index.js:
```diff
@@ -250,6 +340,161 @@ class AuthLdap {
+      xoUser => xoUser.authProviders?.ldap?.id === ldapId
+    )
+    const conflictingXoUser = xoUsers.find(
+      user => user.email === ldapUsername && user.id !== xoUser?.id
```
Wouldn't it be more efficient to stop looking at the first user with the correct name?

julien-f commented on packages/xo-server-auth-ldap/src/index.js:
```diff
@@ -250,6 +340,161 @@ class AuthLdap {
+      if (xoUser === undefined || merge) {
+        xoUser = conflictingXoUser
+      } else {
+        throw new Error('XO user with same username already exists')
```
Add more info to the error to help debugging.

julien-f commented on packages/xo-server-auth-ldap/src/index.js:
```diff
@@ -250,6 +340,161 @@ class AuthLdap {
+    }
+
+    await this._xo.updateUser(xoUser.id, {
+      name: ldapUsername,
```
I'm conflicted about this.

In this case I believe it makes sense but in general, there can multiple providers attached to a user, and providers should not dictate the name of the user.

julien-f replied:
As a work-around, we can first assert there are no other providers for this user.

badrAZ commented on packages/xo-server-auth-ldap/src/index.js:
```diff
@@ -250,6 +340,161 @@ class AuthLdap {
+    )
+
+    if (conflictingXoUser !== undefined) {
+      if (xoUser === undefined || merge) {
```
In case of an existent `xoUser` and an existent `conflictingXOUser` with a `merge` set to `true`, will have a user which refers to the same `ldapUser`. I think in this case we could merge them.

badrAZ commented on packages/xo-server-auth-ldap/src/index.js:
```diff
@@ -238,6 +321,13 @@ class AuthLdap {
           )
           logger(JSON.stringify(entry, null, 2))
+
+          await this._synchronizeUser(username, entry)
```
We only synchronize one user?

pdonias committed:
Don't update user's username if conflicting user was found or user has 2+ auth providers

pdonias committed:
Fix conflicting user handling

pdonias committed:
On user login: only update user's groups

pdonias committed:
Use directory URI as auth provider ID

badrAZ commented on packages/xo-server-auth-ldap/src/index.js:
```diff
@@ -250,6 +348,209 @@ class AuthLdap {
+    // Get the XO user bound to the LDAP user
+    let xoUser = xoUsers.find(
+      xoUser => xoUser.authProviders?.[directoryUri]?.id === ldapId
+    )
```
The user will be synchronized in the most cases after the first synchronization.

IMHO, we can early return when a user is already synchronized (`xoUser?.email === ldapUsername`), it will avoid useless conditions and updates.

badrAZ commented on packages/xo-server-auth-ldap/src/index.js:
```diff
@@ -250,6 +348,209 @@ class AuthLdap {
+          ;[xoGroup] = xoGroups.splice(xoGroupIndex, 1)
+          await this._xo.updateGroup(xoGroup.id, { name: groupLdapName })
+        }
+        xoGroup = await this._xo.getGroup(xoGroup.id)
```
This is needed only on updating  a group, you can move it above.

pdonias committed:
Refetch group only after update

pdonias committed:
Early return XO user

julien-f commented on yarn.lock:
```diff
@@ -11381,17 +11398,17 @@ lcid@^1.0.0:
-  version "2.7.0"
-  resolved "https://registry.yarnpkg.com/ldapts/-/ldapts-2.7.0.tgz#f4fd4c101ea63fa74549695e8f868e54ecf6539e"
-  integrity sha512-eRrokSpUZAN7aRCog0kOPihXdESaMWolnQwURGnKOxkuc19YigjhxRcV/Wjoq5eZc1C8zGSNrV2+q8dj03ppKw==
+  version "2.2.1"
```
Weird changes in your `yarn.lock`, make sure your are up to date with `master`.

pdonias committed:
Merge branch 'master' into pierre-sync-ldap-groups

pdonias committed:
Update yarn.lock

julien-f commented on packages/xo-server/src/utils.js:
```diff
@@ -77,7 +77,7 @@ export const getUserPublicProperties = user =>
     'groups',
     'permission',
     'preferences',
-    'provider'
```
This property was incorrect and unused, it makes sense to remove it.

julien-f commented on packages/xo-server/src/utils.js:
```diff
@@ -77,7 +77,7 @@ export const getUserPublicProperties = user =>
     'permission',
     'preferences',
-    'provider'
+    'authProviders'
```
Keep properties sorted.

julien-f commented on packages/xo-server/src/xo-mixins/subjects.js:
```diff
@@ -139,6 +139,7 @@ export default class {
       password,
       permission,
       preferences,
+      authProviders,
```
Keep properties sorted.

pdonias committed:
Move user creation/update logic to registerUser2

pdonias committed:
Sort

julien-f committed:
Merge branch 'master' into pierre-sync-ldap-groups

julien-f commented on packages/xo-server/src/xo-mixins/subjects.js:
```diff
@@ -231,6 +244,68 @@ export default class {
     })
   }
 
+  async registerUser2(providerId, { user: { id, name }, data }) {
```
Shouldn't we make `registerUser` a small wrapper around this one?

pdonias replied:
What do you mean by that exactly? Making `registerUser` check what the second parameter is before deciding what to do?

julien-f replied:
I mean unifying the code path by making `registerUser` an interface for legacy code which calls `registerUser2` which implement a single logic :slightly_smiling_face:

julien-f commented on packages/xo-server/src/xo-mixins/subjects.js:
```diff
@@ -231,6 +244,68 @@ export default class {
     })
   }
 
+  async registerUser2(providerId, { user: { id, name }, data }) {
```
It would be nice to add a small comment explaining this function?

And maybe what `user.id`, `user.name` and `data` are?

pdonias committed:
Wrap and comment

julien-f commented on packages/xo-server-auth-ldap/src/index.js:
```diff
@@ -172,12 +237,18 @@ class AuthLdap {
     this._searchFilter = searchFilter
     this._startTls = startTls
+    this._groupsConfig = groups
+    this._usersConfig = users
```
Can `groups` (I don't think so) or `users` be undefined`?

If yes, be careful when using them.

julien-f commented on packages/xo-server/src/xo-mixins/subjects.js:
```diff
@@ -212,6 +225,10 @@ export default class {
   // Get or create a user associated with an auth provider.
   async registerUser(provider, name) {
+    if (typeof name === 'object') {
+      return this.registerUser2(provider, name)
```
Not what I was saying, we'll talk about this directly.

julien-f commented on packages/xo-server/src/xo-mixins/subjects.js:
```diff
@@ -231,6 +248,74 @@ export default class {
+      conflictingUser = users.find(user => user.email === name)
+
+      if (conflictingUser !== undefined) {
+        if (!this._xo._config.authentication.mergeProvidersUsers) {
```
Please, document this new settings similarly than: https://github.com/vatesfr/xen-orchestra/blob/ccd34c1610f9eabae3c28adc9b6a552ec45c7e5e/packages/xo-server/config.toml#L45-L46

julien-f commented on packages/xo-server/src/xo-mixins/subjects.js:
```diff
@@ -212,6 +225,10 @@ export default class {
 
   // Get or create a user associated with an auth provider.
```
Add comment to explicit it's deprecated.

julien-f commented on packages/xo-server-auth-ldap/src/index.js:
```diff
@@ -112,6 +116,67 @@ Or something like this if you also want to filter by group:
+          type: 'string',
+        },
+      },
+    },
```
Make sure it works with an old config.

julien-f commented on packages/xo-server-auth-ldap/src/index.js:
```diff
@@ -199,6 +270,10 @@ class AuthLdap {
 
+  _getDirectoryUri() {
+    return new URL(encodeURIComponent(this._searchBase), this._serverUri).href
+  }
```
Don't scope by LDAP server.

pdonias committed:
Multiple fixes

Deprecate registerUser
Unscope LDAP
Default idAttribute to dn
Assign default config values
Document mergeProvidersUsers
Support legacy config

pdonias committed:
idAttribute example: dn

julien-f commented on packages/xo-server/config.toml:
```diff
@@ -42,6 +42,12 @@ perMessageDeflate = { threshold = 524288 } # 512kiB
+# same username is found in XO:
+# - if true: the 2 users will be merged
+# - if false: the user will not be able to log in.
+mergeProvidersUsers = true
```
AFAIU, this behavior is especially useful for the transition period, should we had a comment to remove this in a year?

pdonias replied:
It will also be useful after that, to transition from "users being created manually" to "users being created by the plugin". The plugin will need to know if a native XO user that has the same username as a LDAP user can be considered as the same user.

julien-f commented on packages/xo-server-auth-ldap/src/index.js:
```diff
@@ -112,6 +118,69 @@ Or something like this if you also want to filter by group:
+          description:
+            'Attribute used to identify a user. Must be unique. e.g.: `dn`',
+          type: 'string',
+          default: DEFAULTS.users.idAttribute,
```
These default values are only used for information, and not automatically merged with the config, you need to take care of them yourself.

pdonias replied:
That's the behaviour I'm expecting in the code. e.g.: if the `idAttribute` isn't configured, I'm falling back to the old implementation.

julien-f replied:
Then, IMHO, it's not a default value and should be removed.

pdonias committed:
Add config comment

julien-f commented on packages/xo-server-auth-ldap/src/index.js:
```diff
@@ -244,6 +319,22 @@ class AuthLdap {
+            }
+          }
+
           return { username }
```
The user has already been authenticated by the plugin here, we only need to return the user id.

pdonias committed:
Return userId

julien-f commented on packages/xo-server-auth-ldap/src/index.js:
```diff
@@ -112,6 +118,69 @@ Or something like this if you also want to filter by group:
+        idAttribute: {
+          title: 'ID attribute',
+          description:
+            'Attribute used to identify a group. Must be unique. e.g.: `gid`',
```
I think we should `dn` as a default here.

julien-f commented on packages/xo-server-auth-ldap/src/index.js:
```diff
@@ -112,6 +118,69 @@ Or something like this if you also want to filter by group:
+          description:
+            'Attribute used to find the members of a group. e.g.: `memberUid`. The values must reference the user IDs (cf. user ID attribute)',
+          type: 'string',
+        },
```
I think we need to add a setting for the attribute used for the mapping on the user side.

Maybe with an intermediary object it would be easier to understand:

```
membersMapping: {
  groupAttribute = 'memberUid',
  userAttribute = 'uidNumber',
}
```

pdonias committed:
Members mapping

pdonias committed:
Move group config check

pdonias committed:
Remove prefix and make group config options required

pdonias committed:
userIdAttribute

pdonias committed:
try catch

pdonias committed:
Merge branch 'master' into pierre-sync-ldap-groups

julien-f committed:
Merge branch 'master' into pierre-sync-ldap-groups
